cmd/cerbosctl/hub/store/add_files.go

```diff
@@ -0,0 +1,204 @@
+		}
+
+		stat, err := os.Stat(filePath)
+		if err != nil {
+			return err
+		}
+
+		//nolint:nestif
+		if stat.IsDir() {
+			root, err := os.OpenRoot(filePath)
+			if err != nil {
+				return fmt.Errorf("failed to open %s: %w", filePath, err)
+			}
+			defer root.Close()
+
+			if err := fs.WalkDir(root.FS(), ".", func(path string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+
+				if d.IsDir() {
+					if util.IsHidden(d.Name()) {
+						return fs.SkipDir
+					}
+					return nil
+				}
+
+				if util.IsHidden(d.Name()) {
+					return nil
+				}
+
+				if !util.IsSupportedFileType(d.Name()) {
+					return nil
+				}
+
+				finfo, err := d.Info()
+				if err != nil {
+					return fmt.Errorf("failed to get file information on %s: %w", path, err)
+				}
+
+				if finfo.Size() > maxFileSize {
+					return fmt.Errorf("file too large: %s", path)
+				}
+
+				relativePath, err := filepath.Rel(filePath, path)
+				if err != nil {
+					return fmt.Errorf("failed to determine relative path of %s: %w", path, err)
+				}
+
+				afc.filesToAdd[path] = filepath.Join(storePath, relativePath)
+				return nil
+			}); err != nil {
+				return fmt.Errorf("failed to list files under %s: %w", filePath, err)
+			}
+		} else if util.IsSupportedFileType(filePath) {
+			if stat.Size() > maxFileSize {
+				return fmt.Errorf("file too large: %s", filePath)
+			}
+			afc.filesToAdd[filePath] = storePath
+		}
+	}
+
+	return nil
+}
+
+func (afc *AddFilesCmd) Run(k *kong.Kong, cmd *Cmd) error {
+	client, err := cmd.storeClient()
+	if err != nil {
+		return afc.toCommandError(k.Stderr, err)
+	}
+
+	version := afc.VersionMustEq
+	for batch, err := range afc.batch() {
+		if err != nil {
+			return afc.toCommandError(k.Stderr, err)
+		}
+
+		req := hub.NewModifyFilesRequest(cmd.StoreID, afc.Message).AddOps(batch...)
+		if version > 0 {
+			req.OnlyIfVersionEquals(version)
+		}
+		resp, err := client.ModifyFilesLenient(context.Background(), req)
+		if err != nil {
+			return afc.toCommandError(k.Stderr, err)
+		}
+
+		if resp != nil {
+			version = resp.GetNewStoreVersion()
+		}
+	}
+
+	if version > 0 {
+		afc.printNewVersion(k.Stdout, version)
+		return nil
+	}
+
+	return newStoreNotModifiedError()
+}
+
+func (afc *AddFilesCmd) batch() iter.Seq2[[]*storev1.FileOp, error] {
+	return func(yield func([]*storev1.FileOp, error) bool) {
+		batch := make([]*storev1.FileOp, 0, modifyFilesBatchSize)
+		batchCounter := 0
+
+		for actualPath, storePath := range afc.filesToAdd {
+			contents, err := os.ReadFile(actualPath)
+			if err != nil {
+				yield(nil, fmt.Errorf("failed to read %s: %w", actualPath, err))
+				return
+			}
+
+			if len(contents) > maxFileSize {
+				yield(nil, fmt.Errorf("file too large: %s", actualPath))
+				return
+			}
+
+			batch = append(batch, &storev1.FileOp{
+				Op: &storev1.FileOp_AddOrUpdate{
+					AddOrUpdate: &storev1.File{
+						Path:     storePath,
```

Comment:
storePath here is constructed with filepath.Join, which is OS dependent.

Reply:
I don't think it is? It's either `filepath.Base` of the path or the user-provided value. I've added a `filepath.ToSlash` above as a precaution. 

We don't officially support Windows anyway and there are probably a whole bunch of things that don't work properly on pure Windows.

Reply:
I am going to merge this now because it's needed for a demo. If you spotted anything that needs fixing, please keep commenting and I'll fix them.